astacus/coordinator/plugins/clickhouse/steps.py

```diff
@@ -627,6 +653,22 @@ async def run_step(self, cluster: Cluster, context: StepsContext) -> None:
                     pass
 
 
+@dataclasses.dataclass
+class RestoreUDFsStep(Step[None]):
+    zookeeper_client: ZooKeeperClient
+    replicated_user_defined_zookeeper_path: str | None
+
+    async def run_step(self, cluster: Cluster, context: StepsContext) -> None:
+        if self.replicated_user_defined_zookeeper_path is None:
+            return
+
+        clickhouse_manifest = context.get_result(ClickHouseManifestStep)
+        async with self.zookeeper_client.connect() as connection:
+            for udf in clickhouse_manifest.udfs:
+                path = os.path.join(self.replicated_user_defined_zookeeper_path, udf.path)
```

Comment:
This probably needs a `connection.try_create(self.replicated_user_defined_zookeeper_path)` to make sure the entire path containing all functions exists ?

Reply:
Unneeded

https://github.com/Aiven-Open/astacus/blob/joelynch/backup-zk-udfs/astacus/coordinator/plugins/zookeeper.py#L89

Reply:
Ah right, I forgot that outside of a transaction it's possible/easy to auto-create the entire path :+1: